### PR TITLE
[Old] Added `permutations` and `rotations` iterators

### PR DESCRIPTION
**Pull request by Patitotective**
In `std/algorithm` there is `prev/nextPermutation` and `rotateLeft/Right` but not a simple way of getting all the permutations or all the rotations so this PR adds them both.
~~I also overloaded the iterators to work with `string` directly and not with `openArray[char]` because to convert it back to a string you would need to `join` them which is more expensive.~~

### CODE REVIEW

File: lib/pure/algorithm.nim [hunk 1]
```diff
@@ -755,6 +755,34 @@ proc prevPermutation*[T](x: var openArray[T]): bool {.discardable.} =
...
+    import std/sequtils
+    assert permutations([1, 2, 3]).toSeq == @[@[1, 2, 3], @[1, 3, 2], @[2, 1, 3], @[2, 3, 1], @[3, 1, 2], @[3, 2, 1]]
+
+  var temp = @x
+  temp.sort()
+  
+  yield temp
+
+  while temp.nextPermutation:
+    yield temp
```
Comment (Araq) [inline lib/pure/algorithm.nim]:
```suggestion
  while true:
    yield temp
    if not temp.nextPermutation: break
```

File: lib/pure/algorithm.nim [hunk 2]
```diff
@@ -755,6 +755,34 @@ proc prevPermutation*[T](x: var openArray[T]): bool {.discardable.} =
...
+    import std/sequtils
+    assert permutations("abc").toSeq == @["abc", "acb", "bac", "bca", "cab", "cba"]
+
+  var temp = s
+  temp.sort()
+  
+  yield temp
+
+  while temp.nextPermutation:
+    yield temp
```
Comment (Araq) [inline lib/pure/algorithm.nim]: Likewise.

**Comment by Araq** [comment]
The versions for `string` need to be removed. They don't work for Unicode.

**Comment by Patitotective** [comment]
> The versions for `string` need to be removed. They don't work for Unicode.

Would it be ok to include unicode versions of `permutations` and `rotations` in `std/unicode`?

**Comment by Araq** [comment]
> Would it be ok to include unicode versions of permutations and rotations in std/unicode?

No, add that as a Nimble package instead please. These operations only seem to be useful for code competitions or homework exercises.

**Pull request closed without merging**